core/lib/via_btc_client/src/inscriber/mod.rs

```diff
@@ -1 +1,734 @@
+            signer,
+            context,
+        })
+    }
+
+    // the inscribe should provide report for upper layer to give them information for updates on the transactions
+    // {
+    //    "consumed_utxos": [],
+    //    "commit_tx": {},
+    //    "reveal_tx": {},
+    //    "tx_incldued_in_block": []
+    // }
+    pub async fn inscribe(&mut self, input: lib_types::InscriptionMessage) -> Result<()> {
+        self.sync_context_with_blockchain().await?;
+
+        let secp_ref = &self.signer.get_secp_ref();
+        let internal_key = self.signer.get_internal_key()?;
+        let network = self.client.get_network();
+
+        let inscription_data =
+            InscriptionData::new(input.clone(), secp_ref, internal_key, network)?;
+
+        let commit_tx_input_info = self.prepare_commit_tx_input().await?;
+
+        let commit_tx_output_info = self
+            .prepare_commit_tx_output(
+                &commit_tx_input_info,
+                inscription_data.script_pubkey.clone(),
+            )
+            .await?;
+
+        let final_commit_tx = self.sign_commit_tx(&commit_tx_input_info, &commit_tx_output_info)?;
+
+        let reveal_tx_input_info = self.prepare_reveal_tx_input(
+            &commit_tx_output_info,
+            &final_commit_tx,
+            &inscription_data,
+        )?;
+
+        let reveal_tx_output_info = self
+            .prepare_reveal_tx_output(&reveal_tx_input_info, &inscription_data)
+            .await?;
+
+        let final_reveal_tx = self.sign_reveal_tx(
+            &reveal_tx_input_info,
+            &reveal_tx_output_info,
+            &inscription_data,
+        )?;
+
+        let broadcast_status = self
+            .broadcast_inscription(&final_commit_tx, &final_reveal_tx)
+            .await?;
+
+        if !broadcast_status {
+            return Err(anyhow::anyhow!("Failed to broadcast inscription"));
+        }
+
+        self.insert_inscription_to_context(
+            input,
+            final_commit_tx,
+            final_reveal_tx,
+            commit_tx_output_info,
+            reveal_tx_output_info,
+            commit_tx_input_info,
+        )?;
+
+        Ok(())
+    }
+
+    async fn sync_context_with_blockchain(&mut self) -> Result<()> {
+        if self.context.fifo_queue.is_empty() {
+            return Ok(());
+        }
+
+        let mut new_queue: VecDeque<lib_types::InscriptionRequest> = VecDeque::new();
+
+        let original_queue = self.context.fifo_queue.clone();
+
+        let mut index = 0;
+        while let Some(inscription) = self.context.fifo_queue.pop_front() {
+            let txid_ref = &inscription.fee_payer_ctx.fee_payer_utxo_txid;
+            let res = self
+                .client
+                .check_tx_confirmation(txid_ref, CTX_REQUIRED_CONFIRMATIONS)
+                .await?;
+
+            if index == 0 && !res {
+                // add poped inscription back to the first of queue
+                self.context.fifo_queue = original_queue;
```

Comment:
you are clonning entire queue here. let's not use pop_front, just first or first_mut if you need to change it.
in this way we could get rid of cloning here also

Reply:
in here i used `while`  in my logic and because of that i need pop the first element.

if i want to use `first` it add more operation and time complexity.

and with cloning i could implement it in more readable way and simpler logic. and also after this function end the extra data that we don't need get dropped from memory.
and also in most of time it's not gonna be big queue.

Reply:
Can't we simply use `self.context.fifo_queue.push_front(inscription);` then?

Reply:
```
    async fn sync_context_with_blockchain(&mut self) -> Result<()> {
        if self.context.fifo_queue.is_empty() {
            return Ok(());
        }

        while let Some(inscription) = self.context.fifo_queue.pop_front() {
            let txid_ref = &inscription.fee_payer_ctx.fee_payer_utxo_txid;
            let res = self
                .client
                .check_tx_confirmation(txid_ref, CTX_REQUIRED_CONFIRMATIONS)
                .await?;

            if !res {
                // add popped inscription back to the first of queue and break the loop
                // if the tx is not confirmed yet
                // if the head tx is not confirmed, it means that the rest of the txs are not confirmed
                // so we can break the loop
                self.context.fifo_queue.push_front(inscription);
                break;
            }
        }

        Ok(())
    }
```

i rewrite the whole method like this without using any extra queue and cloning 

thanks guys for correcting me